src/phoenix/datasets/types.py

```diff
@@ -165,48 +226,142 @@ def __post_init__(self) -> None:
-    def annotator_kind(self) -> str: ...
+    def name(self) -> EvaluatorName:
+        if hasattr(self, "_name"):
+            return self._name
+        return self.__class__.__name__
 
+    @property
+    def kind(self) -> EvaluatorKind:
+        if hasattr(self, "_kind"):
+            return self._kind
+        return "CODE"
+
+    def __new__(cls, *args: Any, **kwargs: Any) -> Evaluator:
+        if cls is Evaluator:
+            raise TypeError(f"{cls.__name__} should not be instantiated.")
+        return object.__new__(cls)
 
-@runtime_checkable
-class CanEvaluate(_HasName, _HasKind, Protocol):
     def evaluate(
         self,
-        example: Example,
-        experiment_run: ExperimentRun,
-    ) -> EvaluationResult: ...
-
+        *,
+        output: TaskOutput,
+        expected: ExampleOutput,
+        metadata: ExampleMetadata,
+        input: ExampleInput,
+        **kwargs: Any,
+    ) -> EvaluationResult:
+        raise NotImplementedError
 
-@runtime_checkable
-class CanAsyncEvaluate(_HasName, _HasKind, Protocol):
     async def async_evaluate(
         self,
-        example: Example,
-        experiment_run: ExperimentRun,
-    ) -> EvaluationResult: ...
+        *,
+        output: TaskOutput,
+        expected: ExampleOutput,
+        metadata: ExampleMetadata,
+        input: ExampleInput,
+        **kwargs: Any,
+    ) -> EvaluationResult:
+        return self.evaluate(
+            output=output,
+            expected=expected,
+            metadata=metadata,
+            input=input,
+            **kwargs,
+        )
 
+    def __init_subclass__(cls, is_abstract: bool = False, **kwargs: Any) -> None:
+        super().__init_subclass__(**kwargs)
+        if is_abstract:
+            return
+        evaluate_fn_signature = inspect.signature(Evaluator.evaluate)
+        for super_cls in inspect.getmro(cls):
+            if super_cls in (LLMEvaluator, Evaluator):
+                break
+            has_evaluate = callable(
+                evaluate := super_cls.__dict__.get(Evaluator.evaluate.__name__)
+            ) and validate_evaluate_fn_params(
+                partial(evaluate, None),  # skip first param, i.e. `self`
+                evaluate.__qualname__ if hasattr(evaluate, "__qualname__") else None,
+            )
+            has_async_evaluate = inspect.iscoroutinefunction(
+                async_evaluate := super_cls.__dict__.get(Evaluator.async_evaluate.__name__)
+            ) and validate_evaluate_fn_params(
+                partial(async_evaluate, None),  # skip first param, i.e. `self`
+                async_evaluate.__qualname__ if hasattr(async_evaluate, "__qualname__") else None,
+            )
+            if has_evaluate or has_async_evaluate:
+                return
+        raise EvaluatorImplementationError(
+            f"Evaluator must implement either "
+            f"`def evaluate{evaluate_fn_signature}` or "
+            f"`async def async_evaluate{evaluate_fn_signature}`"
+        )
 
```

Comment:
This is to enforce the existence of the functions because we're not using the abstract method decorator. It also checks that the functions have the required parameters, which ABC doesn't do anyway.

Reply:
What I find confusing is that `ABC` with an `abstractmethod` does not usually support getting the method from a mixin class. So this snippet:

```python
from abc import ABC, abstractmethod


class Animal(ABC):
    @abstractmethod
    def make_noise(self) -> str:
        pass


class Quacks:
    def make_noise(self) -> str:
        return "Quack"


class Duck(Animal, Quacks):
    pass


duck = Duck()
print(f"{duck.make_noise()=}")
```

raises this error

```
Traceback (most recent call last):
  File "/Users/xandersong/Library/Application Support/Code/User/globalStorage/buenon.scratchpads/scratchpads/1a4f2b73c916c9af3926773a674c6453/scratch.py", line 19, in <module>
    duck = Duck()
           ^^^^^^
TypeError: Can't instantiate abstract class Duck without an implementation for abstract method 'make_noise'
```

The `ABC` pattern suggests to me that the implementation of the methods is supposed to live on the subclass (I understand that is not actually the case since we don't have any abstract methods). I think we can nix `ABC` if we want to allow users to use other patterns (e.g., mixins) for defining the required methods.

Reply:
Nice 👍